File: queue_stack/stack.py
class Stack(object):
    class Node():
        def __init__(self, value):
            self.value = value
            self.next = None
    
    class StackEmptyException(Exception):
        pass

    def __init__(self):
        self.top = None

    def pop(self):
        if self.top:
            value = self.top.value
            self.top = self.top.next
            return value
        raise StackEmptyException()

    def push(self, value):
        new = Node(value)
        new.next = self.top
        self.top = new

    def peek(self):
        if self.top:
            return self.top.value
        raise StackEmptyException()

File: CCI/StacksAndQueues/src/Stack/Stack.java
package Stack;

public class Stack<T> {
    public class StackNode<T> {
        public StackNode<T> next;
        public T data = null;

        public StackNode(T data) {
            this.data = data;
        }
    }

    StackNode<T> top;

    public void push(T data) {
        StackNode<T> node = new StackNode(data);
        node.next = top;
        top = node;
    }

    public T pop() {
        if(top != null) {
            T data = top.data;
            top = top.next;
            return data;
        }
        return null;
    }

    public T peek() {
        if(top != null) {
            return top.data;
        }
        return null;
    }

    public static void main(String[] args) {
        Stack<Integer> s = new Stack();

        System.out.println("pop empty: " + s.pop());
        System.out.println("peek empty: " + s.peek());

        s.push(1);
        s.push(2);
        System.out.println("peek 2: " + s.peek());
        System.out.println("pop 2: " + s.pop());
        System.out.println("peek 1: " + s.peek());
        System.out.println("pop 1: " + s.pop());

    }
}
